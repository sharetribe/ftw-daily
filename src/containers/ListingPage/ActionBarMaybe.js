import React from 'react';
import { bool, oneOfType, object } from 'prop-types';
import { FormattedMessage } from '../../util/reactIntl';
import classNames from 'classnames';
import {
  LISTING_STATE_PENDING_APPROVAL,
  LISTING_STATE_CLOSED,
  LISTING_STATE_DRAFT,
  propTypes,
} from '../../util/types';
import { NamedLink } from '../../components';
import EditIcon from './EditIcon';

import css from './ListingPage.module.css';

export const ActionBarMaybe = props => {
  const { isOwnListing, listing, editParams } = props;
  const state = listing.attributes.state;
  const isPendingApproval = state === LISTING_STATE_PENDING_APPROVAL;
  const isClosed = state === LISTING_STATE_CLOSED;
  const isDraft = state === LISTING_STATE_DRAFT;

  if (isOwnListing) {
    let ownListingTextTranslationId = 'ListingPage.ownListing';

    if (isPendingApproval) {
      ownListingTextTranslationId = 'ListingPage.ownListingPendingApproval';
    } else if (isClosed) {
      ownListingTextTranslationId = 'ListingPage.ownClosedListing';
    } else if (isDraft) {
      ownListingTextTranslationId = 'ListingPage.ownListingDraft';
    }

    const message = isDraft ? 'ListingPage.finishListing' : 'ListingPage.editListing';

    const ownListingTextClasses = classNames(css.ownListingText, {
      [css.ownListingTextPendingApproval]: isPendingApproval,
    });

    return (
      <div className={css.actionBar}>
        <p className={ownListingTextClasses}>
          <FormattedMessage id={ownListingTextTranslationId} />
        </p>
        <NamedLink className={css.editListingLink} name="EditListingPage" params={editParams}>
          <EditIcon className={css.editIcon} />
          <FormattedMessage id={message} />
        </NamedLink>
      </div>
    );
  } else if (isClosed) {
    return (
      <div className={css.actionBar}>
        <p className={css.closedListingText}>
          <FormattedMessage id="ListingPage.closedListing" />
        </p>
      </div>
    );
  }
  return null;
};

ActionBarMaybe.propTypes = {
  isOwnListing: bool.isRequired,
  listing: oneOfType([propTypes.listing, propTypes.ownListing]).isRequired,
  editParams: object.isRequired,
};

ActionBarMaybe.displayName = 'ActionBarMaybe';

export default ActionBarMaybe;
