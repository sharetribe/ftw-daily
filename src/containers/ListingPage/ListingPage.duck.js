import pick from 'lodash/pick';
import moment from 'moment';
import config from '../../config';
import { types as sdkTypes } from '../../util/sdkLoader';
import { storableError } from '../../util/errors';
import { addMarketplaceEntities } from '../../ducks/marketplaceData.duck';
import { transactionLineItems } from '../../util/api';
import * as log from '../../util/log';
import { denormalisedResponseEntities } from '../../util/data';
import { TRANSITION_ENQUIRE } from '../../util/transaction';
import {
  LISTING_PAGE_DRAFT_VARIANT,
  LISTING_PAGE_PENDING_APPROVAL_VARIANT,
} from '../../util/urlHelpers';
import { fetchCurrentUser, fetchCurrentUserHasOrdersSuccess } from '../../ducks/user.duck';

const { UUID } = sdkTypes;

// ================ Action types ================ //

export const SET_INITIAL_VALUES = 'app/ListingPage/SET_INITIAL_VALUES';

export const SHOW_LISTING_REQUEST = 'app/ListingPage/SHOW_LISTING_REQUEST';
export const SHOW_LISTING_ERROR = 'app/ListingPage/SHOW_LISTING_ERROR';

export const FETCH_REVIEWS_REQUEST = 'app/ListingPage/FETCH_REVIEWS_REQUEST';
export const FETCH_REVIEWS_SUCCESS = 'app/ListingPage/FETCH_REVIEWS_SUCCESS';
export const FETCH_REVIEWS_ERROR = 'app/ListingPage/FETCH_REVIEWS_ERROR';

export const FETCH_TIME_SLOTS_REQUEST = 'app/ListingPage/FETCH_TIME_SLOTS_REQUEST';
export const FETCH_TIME_SLOTS_SUCCESS = 'app/ListingPage/FETCH_TIME_SLOTS_SUCCESS';
export const FETCH_TIME_SLOTS_ERROR = 'app/ListingPage/FETCH_TIME_SLOTS_ERROR';

export const FETCH_LINE_ITEMS_REQUEST = 'app/ListingPage/FETCH_LINE_ITEMS_REQUEST';
export const FETCH_LINE_ITEMS_SUCCESS = 'app/ListingPage/FETCH_LINE_ITEMS_SUCCESS';
export const FETCH_LINE_ITEMS_ERROR = 'app/ListingPage/FETCH_LINE_ITEMS_ERROR';

export const SEND_ENQUIRY_REQUEST = 'app/ListingPage/SEND_ENQUIRY_REQUEST';
export const SEND_ENQUIRY_SUCCESS = 'app/ListingPage/SEND_ENQUIRY_SUCCESS';
export const SEND_ENQUIRY_ERROR = 'app/ListingPage/SEND_ENQUIRY_ERROR';

// ================ Reducer ================ //

const initialState = {
  id: null,
  showListingError: null,
  reviews: [],
  fetchReviewsError: null,
  timeSlots: null,
  fetchTimeSlotsError: null,
  lineItems: null,
  fetchLineItemsInProgress: false,
  fetchLineItemsError: null,
  sendEnquiryInProgress: false,
  sendEnquiryError: null,
  enquiryModalOpenForListingId: null,
};

const listingPageReducer = (state = initialState, action = {}) => {
  const { type, payload } = action;
  switch (type) {
    case SET_INITIAL_VALUES:
      return { ...initialState, ...payload };

    case SHOW_LISTING_REQUEST:
      return { ...state, id: payload.id, showListingError: null };
    case SHOW_LISTING_ERROR:
      return { ...state, showListingError: payload };

    case FETCH_REVIEWS_REQUEST:
      return { ...state, fetchReviewsError: null };
    case FETCH_REVIEWS_SUCCESS:
      return { ...state, reviews: payload };
    case FETCH_REVIEWS_ERROR:
      return { ...state, fetchReviewsError: payload };

    case FETCH_TIME_SLOTS_REQUEST:
      return { ...state, fetchTimeSlotsError: null };
    case FETCH_TIME_SLOTS_SUCCESS:
      return { ...state, timeSlots: payload };
    case FETCH_TIME_SLOTS_ERROR:
      return { ...state, fetchTimeSlotsError: payload };

    case FETCH_LINE_ITEMS_REQUEST:
      return { ...state, fetchLineItemsInProgress: true, fetchLineItemsError: null };
    case FETCH_LINE_ITEMS_SUCCESS:
      return { ...state, fetchLineItemsInProgress: false, lineItems: payload };
    case FETCH_LINE_ITEMS_ERROR:
      return { ...state, fetchLineItemsInProgress: false, fetchLineItemsError: payload };

    case SEND_ENQUIRY_REQUEST:
      return { ...state, sendEnquiryInProgress: true, sendEnquiryError: null };
    case SEND_ENQUIRY_SUCCESS:
      return { ...state, sendEnquiryInProgress: false };
    case SEND_ENQUIRY_ERROR:
      return { ...state, sendEnquiryInProgress: false, sendEnquiryError: payload };

    default:
      return state;
  }
};

export default listingPageReducer;

// ================ Action creators ================ //

export const setInitialValues = initialValues => ({
  type: SET_INITIAL_VALUES,
  payload: pick(initialValues, Object.keys(initialState)),
});

export const showListingRequest = id => ({
  type: SHOW_LISTING_REQUEST,
  payload: { id },
});

export const showListingError = e => ({
  type: SHOW_LISTING_ERROR,
  error: true,
  payload: e,
});

export const fetchReviewsRequest = () => ({ type: FETCH_REVIEWS_REQUEST });
export const fetchReviewsSuccess = reviews => ({ type: FETCH_REVIEWS_SUCCESS, payload: reviews });
export const fetchReviewsError = error => ({
  type: FETCH_REVIEWS_ERROR,
  error: true,
  payload: error,
});

export const fetchTimeSlotsRequest = () => ({ type: FETCH_TIME_SLOTS_REQUEST });
export const fetchTimeSlotsSuccess = timeSlots => ({
  type: FETCH_TIME_SLOTS_SUCCESS,
  payload: timeSlots,
});
export const fetchTimeSlotsError = error => ({
  type: FETCH_TIME_SLOTS_ERROR,
  error: true,
  payload: error,
});

export const fetchLineItemsRequest = () => ({ type: FETCH_LINE_ITEMS_REQUEST });
export const fetchLineItemsSuccess = lineItems => ({
  type: FETCH_LINE_ITEMS_SUCCESS,
  payload: lineItems,
});
export const fetchLineItemsError = error => ({
  type: FETCH_LINE_ITEMS_ERROR,
  error: true,
  payload: error,
});

export const sendEnquiryRequest = () => ({ type: SEND_ENQUIRY_REQUEST });
export const sendEnquirySuccess = () => ({ type: SEND_ENQUIRY_SUCCESS });
export const sendEnquiryError = e => ({ type: SEND_ENQUIRY_ERROR, error: true, payload: e });

// ================ Thunks ================ //

export const showListing = (listingId, isOwn = false) => (dispatch, getState, sdk) => {
  dispatch(showListingRequest(listingId));
  dispatch(fetchCurrentUser());
  const params = {
    id: listingId,
    include: ['author', 'author.profileImage', 'images'],
    'fields.image': [
      // Listing page
      'variants.landscape-crop',
      'variants.landscape-crop2x',
      'variants.landscape-crop4x',
      'variants.landscape-crop6x',

      // Social media
      'variants.facebook',
      'variants.twitter',

      // Image carousel
      'variants.scaled-small',
      'variants.scaled-medium',
      'variants.scaled-large',
      'variants.scaled-xlarge',

      // Avatars
      'variants.square-small',
      'variants.square-small2x',
    ],
  };

  const show = isOwn ? sdk.ownListings.show(params) : sdk.listings.show(params);

  return show
    .then(data => {
      dispatch(addMarketplaceEntities(data));
      return data;
    })
    .catch(e => {
      dispatch(showListingError(storableError(e)));
    });
};

export const fetchReviews = listingId => (dispatch, getState, sdk) => {
  dispatch(fetchReviewsRequest());
  return sdk.reviews
    .query({
      listing_id: listingId,
      state: 'public',
      include: ['author', 'author.profileImage'],
      'fields.image': ['variants.square-small', 'variants.square-small2x'],
    })
    .then(response => {
      const reviews = denormalisedResponseEntities(response);
      dispatch(fetchReviewsSuccess(reviews));
    })
    .catch(e => {
      dispatch(fetchReviewsError(storableError(e)));
    });
};

const timeSlotsRequest = params => (dispatch, getState, sdk) => {
  return sdk.timeslots.query(params).then(response => {
    return denormalisedResponseEntities(response);
  });
};

export const fetchTimeSlots = listingId => (dispatch, getState, sdk) => {
  dispatch(fetchTimeSlotsRequest);

  // Time slots can be fetched for 90 days at a time,
  // for at most 180 days from now. If max number of bookable
  // day exceeds 90, a second request is made.

  const maxTimeSlots = 90;
  // booking range: today + bookable days -1
  const bookingRange = config.dayCountAvailableForBooking - 1;
  const timeSlotsRange = Math.min(bookingRange, maxTimeSlots);

  const start = moment
    .utc()
    .startOf('day')
    .toDate();
  const end = moment()
    .utc()
    .startOf('day')
    .add(timeSlotsRange, 'days')
    .toDate();
  const params = { listingId, start, end };

  return dispatch(timeSlotsRequest(params))
    .then(timeSlots => {
      const secondRequest = bookingRange > maxTimeSlots;

      if (secondRequest) {
        const secondRange = Math.min(maxTimeSlots, bookingRange - maxTimeSlots);
        const secondParams = {
          listingId,
          start: end,
          end: moment(end)
            .add(secondRange, 'days')
            .toDate(),
        };

        return dispatch(timeSlotsRequest(secondParams)).then(secondBatch => {
          const combined = timeSlots.concat(secondBatch);
          dispatch(fetchTimeSlotsSuccess(combined));
        });
      } else {
        dispatch(fetchTimeSlotsSuccess(timeSlots));
      }
    })
    .catch(e => {
      dispatch(fetchTimeSlotsError(storableError(e)));
    });
};

export const sendEnquiry = (listingId, message) => (dispatch, getState, sdk) => {
  dispatch(sendEnquiryRequest());
  const bodyParams = {
    transition: TRANSITION_ENQUIRE,
    processAlias: config.bookingProcessAlias,
    params: { listingId },
  };
  return sdk.transactions
    .initiate(bodyParams)
    .then(response => {
      const transactionId = response.data.data.id;

      // Send the message to the created transaction
      return sdk.messages.send({ transactionId, content: message }).then(() => {
        dispatch(sendEnquirySuccess());
        dispatch(fetchCurrentUserHasOrdersSuccess(true));
        return transactionId;
      });
    })
    .catch(e => {
      dispatch(sendEnquiryError(storableError(e)));
      throw e;
    });
};

export const fetchTransactionLineItems = ({ bookingData, listingId, isOwnListing }) => dispatch => {
  dispatch(fetchLineItemsRequest());
  transactionLineItems({ bookingData, listingId, isOwnListing })
    .then(response => {
      const lineItems = response.data;
      dispatch(fetchLineItemsSuccess(lineItems));
    })
    .catch(e => {
      dispatch(fetchLineItemsError(storableError(e)));
      log.error(e, 'fetching-line-items-failed', {
        listingId: listingId.uuid,
        bookingData: bookingData,
      });
    });
};

export const loadData = (params, search) => dispatch => {
  const listingId = new UUID(params.id);

  const ownListingVariants = [LISTING_PAGE_DRAFT_VARIANT, LISTING_PAGE_PENDING_APPROVAL_VARIANT];
  if (ownListingVariants.includes(params.variant)) {
    return dispatch(showListing(listingId, true));
  }

  if (config.enableAvailability) {
    return Promise.all([
      dispatch(showListing(listingId)),
      dispatch(fetchTimeSlots(listingId)),
      dispatch(fetchReviews(listingId)),
    ]);
  } else {
    return Promise.all([dispatch(showListing(listingId)), dispatch(fetchReviews(listingId))]);
  }
};
