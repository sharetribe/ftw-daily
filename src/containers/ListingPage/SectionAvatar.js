import React from 'react';
import { NamedLink, AvatarLarge, AvatarMedium } from '../../components';

import css from './ListingPage.module.css';

const SectionAvatar = props => {
  const { user, params } = props;
  return (
    <div className={css.sectionAvatar}>
      <NamedLink name="ListingPage" params={params} to={{ hash: '#host' }}>
        <AvatarLarge user={user} className={css.avatarDesktop} disableProfileLink />
      </NamedLink>
      <NamedLink name="ListingPage" params={params} to={{ hash: '#host' }}>
        <AvatarMedium user={user} className={css.avatarMobile} disableProfileLink />
      </NamedLink>
    </div>
  );
};

export default SectionAvatar;
