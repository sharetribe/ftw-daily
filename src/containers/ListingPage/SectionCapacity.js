import React from 'react';
import { array, shape, string } from 'prop-types';
import { FormattedMessage } from 'react-intl';

import css from './ListingPage.module.css';

const SectionCapacity = props => {
  const { publicData, options } = props;

  const capacity = publicData.capacity;
  const capacityOption = options.find(
    option => option.key === capacity
  );

  return capacityOption ? (
    <div className={css.sectionCapacity}>
      <h2 className={css.capacityTitle}>
        <FormattedMessage id="ListingPage.capacityTitle" />
      </h2>
      <p className={css.capacity}>{capacityOption.label}</p>
    </div>
  ) : null;
};

SectionCapacity.propTypes = {
  options: array.isRequired,
  publicData: shape({
    capacity: string,
  }).isRequired,
};

export default SectionCapacity;
