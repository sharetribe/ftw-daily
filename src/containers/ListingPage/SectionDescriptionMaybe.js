import React from 'react';
import { FormattedMessage } from '../../util/reactIntl';
import { richText } from '../../util/richText';

import css from './ListingPage.module.css';

const MIN_LENGTH_FOR_LONG_WORDS_IN_DESCRIPTION = 20;

const SectionDescriptionMaybe = props => {
  const { description } = props;
  return description ? (
    <div className={css.sectionDescription}>
      <h2 className={css.descriptionTitle}>
        <FormattedMessage id="ListingPage.descriptionTitle" />
      </h2>
      <p className={css.description}>
        {richText(description, {
          longWordMinLength: MIN_LENGTH_FOR_LONG_WORDS_IN_DESCRIPTION,
          longWordClass: css.longWord,
        })}
      </p>
    </div>
  ) : null;
};

export default SectionDescriptionMaybe;
