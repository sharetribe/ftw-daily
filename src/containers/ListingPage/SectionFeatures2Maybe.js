import React from 'react';
import { FormattedMessage } from '../../util/reactIntl';
import { PropertyGroup } from '../../components';

import css from './ListingPage.module.css';

const SectionFeaturesMaybe = props => {
  const { options, publicData ,hostLink } = props;
  if (!publicData) {
    return null;
  }

  const selectedOptions = publicData && publicData.typeOfPets ? publicData.numberOfPets : [];
  return (
    <div className={css.sectionFeatures}>
      <h2 className={css.featuresTitle}>
        <FormattedMessage id="ListingPage.featuresnumberpet" values={{ name: hostLink }}  />
      </h2>
      <PropertyGroup
        id="ListingPage.typeOfPets"
        options={options}
        selectedOptions={selectedOptions}
        twoColumns={true}
      />
    </div>
  );
};

export default SectionFeaturesMaybe;
