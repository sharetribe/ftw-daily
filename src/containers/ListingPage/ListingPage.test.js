import React from 'react';
import { shallow } from 'enzyme';
import { FormattedMessage } from '../../util/reactIntl';
import { types as sdkTypes } from '../../util/sdkLoader';
import {
  createUser,
  createCurrentUser,
  createListing,
  createOwnListing,
  fakeIntl,
} from '../../util/test-data';
import { storableError } from '../../util/errors';
import { renderShallow } from '../../util/test-helpers';
import {
  LINE_ITEM_NIGHT,
  LISTING_STATE_PENDING_APPROVAL,
  LISTING_STATE_PUBLISHED,
  LISTING_STATE_CLOSED,
} from '../../util/types';
import { addMarketplaceEntities } from '../../ducks/marketplaceData.duck';
import { showListingRequest, showListingError, showListing } from './ListingPage.duck';

import { ListingPageComponent } from './ListingPage';
import ActionBarMaybe from './ActionBarMaybe';

const { UUID } = sdkTypes;
const noop = () => null;

const filterConfig = [
  {
    id: 'category',
    label: 'Category',
    type: 'SelectSingleFilter',
    group: 'secondary',
    queryParamName: 'pub_category',
    config: {
      options: [{ key: 'cat1', label: 'Cat 1' }, { key: 'cat2', label: 'Cat 2' }],
    },
  },
  {
    id: 'amenities',
    label: 'Amenities',
    type: 'SelectMultipleFilter',
    group: 'secondary',
    queryParamName: 'pub_amenities',
    config: {
      mode: 'has_all',
      options: [
        {
          key: 'feat1',
          label: 'Feat 1',
        },
        {
          key: 'feat2',
          label: 'Feat 2',
        },
        {
          key: 'feat3',
          label: 'Feat 3',
        },
      ],
    },
  },
];
describe('ListingPage', () => {
  it('matches snapshot', () => {
    const currentUser = createCurrentUser('user-2');
    const id = 'listing1';
    const slug = 'listing1-title';
    const listing1 = createListing(id, {}, { author: createUser('user-1') });
    const listing1Own = createOwnListing(id, {}, { author: createCurrentUser('user-1') });
    const getListing = () => listing1;
    const getOwnListing = () => listing1Own;

    const props = {
      unitType: LINE_ITEM_NIGHT,
      location: {
        pathname: `/l/${slug}/${id}`,
        search: '',
        hash: '',
      },
      history: {
        push: () => console.log('HistoryPush called'),
      },
      params: { id, slug },
      currentUser,
      getListing,
      getOwnListing,
      intl: fakeIntl,
      authInProgress: false,
      currentUserHasListings: false,
      isAuthenticated: false,
      onLogout: noop,
      onLoadListing: noop,
      onManageDisableScrolling: noop,
      scrollingDisabled: false,
      callSetInitialValues: noop,
      sendVerificationEmailInProgress: false,
      onResendVerificationEmail: noop,
      onInitializeCardPaymentData: noop,
      sendEnquiryInProgress: false,
      onSendEnquiry: noop,
      filterConfig,
      fetchLineItemsInProgress: false,
      onFetchTransactionLineItems: () => null,
    };

    const tree = renderShallow(<ListingPageComponent {...props} />);
    expect(tree).toMatchSnapshot();
  });

  describe('Duck', () => {
    it('showListing() success', () => {
      const id = new UUID('00000000-0000-0000-0000-000000000000');
      const dispatch = jest.fn(action => action);
      const response = { status: 200 };
      const show = jest.fn(() => Promise.resolve(response));
      const sdk = { listings: { show }, currentUser: { show } };

      return showListing(id)(dispatch, null, sdk).then(data => {
        expect(data).toEqual(response);
        expect(show.mock.calls).toEqual([
          [
            {
              id,
              include: ['author', 'author.profileImage', 'images'],
              'fields.image': [
                // Listing page
                'variants.landscape-crop',
                'variants.landscape-crop2x',
                'variants.landscape-crop4x',
                'variants.landscape-crop6x',

                // Social media
                'variants.facebook',
                'variants.twitter',

                // Image carousel
                'variants.scaled-small',
                'variants.scaled-medium',
                'variants.scaled-large',
                'variants.scaled-xlarge',

                // Avatars
                'variants.square-small',
                'variants.square-small2x',
              ],
            },
          ],
        ]);
        expect(dispatch.mock.calls).toEqual([
          [showListingRequest(id)],
          [expect.anything()], // fetchCurrentUser() call
          [addMarketplaceEntities(data)],
        ]);
      });
    });

    it('showListing() error', () => {
      const id = new UUID('00000000-0000-0000-0000-000000000000');
      const dispatch = jest.fn(action => action);
      const error = new Error('fail');
      const show = jest.fn(() => Promise.reject(error));
      const sdk = { listings: { show } };

      // Calling sdk.listings.show is expected to fail now

      return showListing(id)(dispatch, null, sdk).then(data => {
        expect(show.mock.calls).toEqual([
          [
            {
              id,
              include: ['author', 'author.profileImage', 'images'],
              'fields.image': [
                // Listing page
                'variants.landscape-crop',
                'variants.landscape-crop2x',
                'variants.landscape-crop4x',
                'variants.landscape-crop6x',

                // Social media
                'variants.facebook',
                'variants.twitter',

                // Image carousel
                'variants.scaled-small',
                'variants.scaled-medium',
                'variants.scaled-large',
                'variants.scaled-xlarge',

                // Avatars
                'variants.square-small',
                'variants.square-small2x',
              ],
            },
          ],
        ]);
        expect(dispatch.mock.calls).toEqual([
          [showListingRequest(id)],
          [expect.anything()], // fetchCurrentUser() call
          [showListingError(storableError(error))],
        ]);
      });
    });
  });

  describe('ActionBarMaybe', () => {
    it('shows users own listing status', () => {
      const listing = createListing('listing-published', {
        state: LISTING_STATE_PUBLISHED,
      });
      const actionBar = shallow(<ActionBarMaybe isOwnListing listing={listing} editParams={{}} />);
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(2);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.ownListing');
      expect(formattedMessages.at(1).props().id).toEqual('ListingPage.editListing');
    });
    it('shows users own pending listing status', () => {
      const listing = createListing('listing-published', {
        state: LISTING_STATE_PENDING_APPROVAL,
      });
      const actionBar = shallow(<ActionBarMaybe isOwnListing listing={listing} editParams={{}} />);
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(2);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.ownListingPendingApproval');
      expect(formattedMessages.at(1).props().id).toEqual('ListingPage.editListing');
    });
    it('shows users own closed listing status', () => {
      const listing = createListing('listing-closed', {
        state: LISTING_STATE_CLOSED,
      });
      const actionBar = shallow(<ActionBarMaybe isOwnListing listing={listing} editParams={{}} />);
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(2);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.ownClosedListing');
      expect(formattedMessages.at(1).props().id).toEqual('ListingPage.editListing');
    });
    it('shows closed listing status', () => {
      const listing = createListing('listing-closed', {
        state: LISTING_STATE_CLOSED,
      });
      const actionBar = shallow(
        <ActionBarMaybe isOwnListing={false} listing={listing} editParams={{}} />
      );
      const formattedMessages = actionBar.find(FormattedMessage);
      expect(formattedMessages.length).toEqual(1);
      expect(formattedMessages.at(0).props().id).toEqual('ListingPage.closedListing');
    });
    it("is missing if listing is not closed or user's own", () => {
      const listing = createListing('listing-published', {
        state: LISTING_STATE_PUBLISHED,
      });
      const actionBar = shallow(
        <ActionBarMaybe isOwnListing={false} listing={listing} editParams={{}} />
      );
      expect(actionBar.getElement()).toBeNull();
    });
  });
});
