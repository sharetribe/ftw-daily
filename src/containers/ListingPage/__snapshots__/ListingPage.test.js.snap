// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingPage matches snapshot 1`] = `
<Page
  author="user-1 display name"
  contentType="website"
  description="listing1 description"
  facebookImages={Array []}
  schema={
    Object {
      "@context": "http://schema.org",
      "@type": "ItemPage",
      "description": "listing1 description",
      "image": "[]",
      "name": "ListingPage.schemaTitle",
    }
  }
  scrollingDisabled={false}
  title="ListingPage.schemaTitle"
  twitterImages={Array []}
>
  <LayoutSingleColumn
    className="pageRoot"
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent)) />
    </LayoutWrapperTopbar>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <div>
        <SectionImages
          editParams={
            Object {
              "id": "listing1",
              "slug": "listing1-title",
              "tab": "description",
              "type": "edit",
            }
          }
          handleViewPhotosClick={[Function]}
          imageCarouselOpen={false}
          isOwnListing={false}
          listing={
            Object {
              "attributes": Object {
                "deleted": false,
                "description": "listing1 description",
                "geolocation": LatLng {
                  "_sdkType": "LatLng",
                  "lat": 40,
                  "lng": 60,
                },
                "price": Money {
                  "_sdkType": "Money",
                  "amount": 5500,
                  "currency": "USD",
                },
                "publicData": Object {},
                "state": "published",
                "title": "listing1 title",
              },
              "author": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "user-1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "user-1",
                },
                "type": "user",
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "listing1",
              },
              "images": Array [],
              "type": "listing",
            }
          }
          onImageCarouselClose={[Function]}
          onManageDisableScrolling={[Function]}
          title="listing1 title"
        />
        <div
          className="contentContainer"
        >
          <SectionAvatar
            params={
              Object {
                "id": "listing1",
                "slug": "listing1-title",
              }
            }
            user={
              Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "user-1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "user-1",
                },
                "type": "user",
              }
            }
          />
          <div
            className="mainContent"
          >
            <SectionHeading
              category={null}
              formattedPrice={55}
              hostLink={
                <NamedLink
                  className="authorNameLink"
                  name="ListingPage"
                  params={
                    Object {
                      "id": "listing1",
                      "slug": "listing1-title",
                    }
                  }
                  to={
                    Object {
                      "hash": "#host",
                    }
                  }
                >
                  user-1 display name
                </NamedLink>
              }
              onContactUser={[Function]}
              priceTitle={55}
              richTitle={
                <span>
                  
                  listing1
                   
                  title
                  
                </span>
              }
              showContactUser={true}
            />
            <SectionDescriptionMaybe
              description="listing1 description"
            />
            <SectionFeaturesMaybe
              options={
                Array [
                  Object {
                    "key": "feat1",
                    "label": "Feat 1",
                  },
                  Object {
                    "key": "feat2",
                    "label": "Feat 2",
                  },
                  Object {
                    "key": "feat3",
                    "label": "Feat 3",
                  },
                ]
              }
              publicData={Object {}}
            />
            <SectionRulesMaybe
              className={null}
              publicData={Object {}}
              rootClassName={null}
            />
            <SectionMapMaybe
              className={null}
              geolocation={
                LatLng {
                  "_sdkType": "LatLng",
                  "lat": 40,
                  "lng": 60,
                }
              }
              listingId={
                UUID {
                  "_sdkType": "UUID",
                  "uuid": "listing1",
                }
              }
              publicData={Object {}}
              rootClassName={null}
            />
            <SectionReviews
              fetchReviewsError={null}
              reviews={Array []}
            />
            <SectionHostMaybe
              authorDisplayName="user-1 display name"
              currentUser={
                Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "email": "user-2@example.com",
                    "emailVerified": true,
                    "profile": Object {
                      "abbreviatedName": "user-2 abbreviated name",
                      "displayName": "user-2 display name",
                      "firstName": "user-2 first name",
                      "lastName": "user-2 last name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "user-2",
                  },
                  "type": "currentUser",
                }
              }
              isEnquiryModalOpen={false}
              listing={
                Object {
                  "attributes": Object {
                    "deleted": false,
                    "description": "listing1 description",
                    "geolocation": LatLng {
                      "_sdkType": "LatLng",
                      "lat": 40,
                      "lng": 60,
                    },
                    "price": Money {
                      "_sdkType": "Money",
                      "amount": 5500,
                      "currency": "USD",
                    },
                    "publicData": Object {},
                    "state": "published",
                    "title": "listing1 title",
                  },
                  "author": Object {
                    "attributes": Object {
                      "banned": false,
                      "deleted": false,
                      "profile": Object {
                        "abbreviatedName": "TT",
                        "displayName": "user-1 display name",
                      },
                    },
                    "id": UUID {
                      "_sdkType": "UUID",
                      "uuid": "user-1",
                    },
                    "type": "user",
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "listing1",
                  },
                  "images": Array [],
                  "type": "listing",
                }
              }
              onCloseEnquiryModal={[Function]}
              onContactUser={[Function]}
              onManageDisableScrolling={[Function]}
              onSubmitEnquiry={[Function]}
              sendEnquiryError={null}
              sendEnquiryInProgress={false}
              title="listing1 title"
            />
          </div>
          <withRouter(injectIntl(BookingPanel))
            authorDisplayName="user-1 display name"
            className="bookingPanel"
            fetchLineItemsError={null}
            fetchLineItemsInProgress={false}
            fetchTimeSlotsError={null}
            isOwnListing={false}
            lineItems={null}
            listing={
              Object {
                "attributes": Object {
                  "deleted": false,
                  "description": "listing1 description",
                  "geolocation": LatLng {
                    "_sdkType": "LatLng",
                    "lat": 40,
                    "lng": 60,
                  },
                  "price": Money {
                    "_sdkType": "Money",
                    "amount": 5500,
                    "currency": "USD",
                  },
                  "publicData": Object {},
                  "state": "published",
                  "title": "listing1 title",
                },
                "author": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "user-1 display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "user-1",
                  },
                  "type": "user",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "listing1",
                },
                "images": Array [],
                "type": "listing",
              }
            }
            onFetchTransactionLineItems={[Function]}
            onManageDisableScrolling={[Function]}
            onSubmit={[Function]}
            subTitle="ListingPage.bookingSubTitle"
            timeSlots={null}
            title={
              <Memo(MemoizedFormattedMessage)
                id="ListingPage.bookingTitle"
                values={
                  Object {
                    "title": <span>
                      
                      listing1
                       
                      title
                      
                    </span>,
                  }
                }
              />
            }
            unitType="line-item/night"
          />
        </div>
      </div>
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className={null}
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSingleColumn>
</Page>
`;
